{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Deliverables:  40 total points\n",
    "\n",
    " - this notebook with inspection intro and results filled out for CoordinateMathSoln  4 points\n",
    " - seqCleaner  9 points\n",
    " - fastqParse  9 points\n",
    " - coordinateMathSoln 9 points\n",
    " - converter 9 points\n",
    " - Optional: Triad class rewrite (5 points extra credit)\n",
    " \n",
    " These must be submitted as the notebook and 4(or 5) .py files, \n",
    "     \n",
    " Due: Monday January 17, 2022 11:55pm\n",
    "\n",
    "# Lab 2 – Manipulating Data Types (40 points)\n",
    "## Overview\n",
    "Last week we introduced the Python function input(), which is used to take in string data. This week we will use input() to take in string and numeric (float and integer) data. \n",
    "\n",
    "Remember that all data returned from input() is a string object, so this will mean that you need to convert any numeric data to their respective numeric objects. The exercises in this lab will give you practice manipulating various types of data that commonly arise in computational biology problems.\n",
    "\n",
    "We are again using jupyter for this assignment, and as a final step we will create distinct command line programs from each of the parts.\n",
    "Please save each of the four (or five) python programs as .py files into your LAB02 folder. You can copy the text from the code cell, then paste it into an editor like notepad or textedit. Save that new file with the appropriate name ( eg. seqCleaner.py) into your LAB02 folder. Using terminal or cmd, you can now do a final test on that new program with the command:<br>\n",
    "python3 seqCleaner <br>\n",
    "\n",
    "All together, you will submit:\n",
    " - this notebook, with the inspection sections filled out\n",
    " - seqCleaner.py, \n",
    " - fastqParse,py, \n",
    " - coordinateMathSoln.py, \n",
    " - converter.py, and \n",
    " - (optionally) Triad.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence cleanup\n",
    "In this exercise, you will create a program to “clean up” a sequence of DNA by removing ambiguous bases (denoted by “N”) output from a sequencer. Your task is to create a Python program called seqCleaner that\n",
    " - asks for and collects a sequence of DNA using input()\n",
    " - removes the ambiguous parts of the sequence, outputs the “cleaned” sequence, replacing the ambiguous parts with a count in {}’s.\n",
    "For example, if I enter the sequence of DNA “AaNNNNNNGTC” (without quotes), the program will output:<br>\n",
    "AA{6}GTC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hints: \n",
    " - The input sequence is not guaranteed to be uppercase, but should be interpreted as though it is all uppercase.\n",
    " - Only the letters (A,C,G,T,N) will be included in the input.\n",
    " - Only the DNA characters (A,C,G,T) should remain after the cleanup. \n",
    " - The input may include, at most, one block of 'N' characters. \n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and specific line or block comments that explain the semantics of your implementation.\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs. This should be in the form of a  program level dosctring."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### seqCleaner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA data?AaGTC\n",
      "AAGTC\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Cindy Liang \n",
    "# Group Members: Ben Matera\n",
    "\n",
    "'''\n",
    "Program overview\n",
    "Read a DNA string from user input and return a collapsed substring of embedded Ns to: {count}.\n",
    "\n",
    "Example: \n",
    " input: AaNNNNNNGTC\n",
    "output: AA{6}GTC\n",
    "\n",
    "Any lower case letters are converted to uppercase\n",
    "\n",
    "Assumptions: \n",
    "Only the letters (A,C,G,T,N) will be included in the input.\n",
    "The input may include, at most, one block of 'N' characters.\n",
    "\n",
    "'''\n",
    "\n",
    "#create class dnastring that takes input and returns the length and the cleaned sequence\n",
    "\n",
    "'''Create a DNAstring class that checks if input string has Ns'''\n",
    "\n",
    "class DNAstring (str):\n",
    "    def length (self):\n",
    "        return (length(self))\n",
    "  \n",
    "    def purify(self):\n",
    "        '''Replace N with number of Ns if Ns are present'''\n",
    "        uprCase = self.upper() \n",
    "        # assume only one N block will appear at a time and count Ns\n",
    "        if 'N' in uprCase:\n",
    "            numN = uprCase.count('N')\n",
    "            # replace N block with number of Ns\n",
    "            newDNA = uprCase.replace('N'*numN, \"{\"+str(numN)+\"}\")\n",
    "            return newDNA\n",
    "        #if no Ns, just return capitalized sequence\n",
    "        else:\n",
    "            return uprCase\n",
    "            \n",
    "def main():\n",
    "    ''' Get user DNA data and clean it up.'''\n",
    "    data = input('DNA data?: ')\n",
    "    thisDNA = DNAstring (data)\n",
    "    pureData = thisDNA.purify()\n",
    "    print (pureData)\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence information parsing\n",
    "In this exercise, you will create a program to “parse” sequence name information from a single line of a FASTQ formatted file. Your task is to create a Python script called fastqParse that:\n",
    " - asks for and collects the seqname line of a FASTQ file using input()\n",
    " - parses out each field of the run information from the string and displays each of them on a new line\n",
    "For example, if I enter the FASTQ seqname line:<br> \n",
    "__*@EAS139:136:FC706VJ:2:2104:15343:197393*__\n",
    "then the program will output:<br>\n",
    "Instrument = EAS139<br>\n",
    "Run ID = 136<br>\n",
    "Flow Cell ID = FC706VJ<br>\n",
    "Flow Cell Lane = 2<br>\n",
    "Tile Number = 2104<br>\n",
    "X-coord = 15343<br>\n",
    "Y-coord = 197393\n",
    "### Hints: \n",
    " - The input string is guaranteed to have 7 fields. \n",
    " - The first character of the FASTQ seqname line is “@” and each field of the run information is separated by a colon“:”.  \n",
    " - A reasonable solution would be around 16 lines of code excluding comments.\n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain documentation/comments\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### fastqParse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter FastQ seqname line@EAS139:136:FC706VJ:2:2104:15343:197393\n",
      "Instrument = EAS139\n",
      "Run ID = 136\n",
      "Flow Cell ID = FC706VJ\n",
      "Flow Cell lane = 2\n",
      "Tile Number = 2104\n",
      "X-coord = 15343\n",
      "Y-coord = 197393\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: YCindy Liang\n",
    "# Group Members: Ben Matera\n",
    "\n",
    "'''\n",
    "Program Overview\n",
    "\n",
    "Input: string containing FASTQ seqname line\n",
    "Output: Each field of sequence run information on a new line\n",
    "\n",
    "Assumptions: \n",
    "The input string is guaranteed to have 7 fields\n",
    "The first character of the FASTQ seqname line is “@” and each field of the run information is separated by a colon“:”\n",
    "A reasonable solution would be around 16 lines of code excluding comments.\n",
    "\n",
    "'''\n",
    "\n",
    "class FastqString (str):\n",
    "    ''' Initiate FastqString class to collect and identify sequence run fields from input.'''\n",
    "    def parse(self):\n",
    "        ''' Split input by : to create a list of properties. Assign each property to the field it belongs to and print results'''\n",
    "       \n",
    "        #split input string by :         \n",
    "        myList = self.split(':')\n",
    "       \n",
    "        #get instrument, flow ID, other information from input\n",
    "        instrument = myList[0]\n",
    "        self.runID = myList[1]\n",
    "        self.flowID = myList[2]\n",
    "        self.flowLane = myList[3]\n",
    "        self.tileNum = myList[4]\n",
    "        self.xCoord = myList[5]\n",
    "        self.yCoord = myList[6]\n",
    "        \n",
    "        #clean up instrument field\n",
    "        self.insID = instrument.replace(\"@\", \"\")\n",
    "\n",
    "        #print information to screen\n",
    "        print(\"Instrument = {0}\".format (self.insID))\n",
    "        print(\"Run ID = {0}\".format (self.runID))\n",
    "        print(\"Flow Cell ID = {0}\".format (self.flowID))\n",
    "        print(\"Flow Cell lane = {0}\".format (self.flowLane))\n",
    "        print(\"Tile Number = {0}\".format (self.tileNum))\n",
    "        print(\"X-coord = {0}\".format (self.xCoord))\n",
    "        print(\"Y-coord = {0}\".format (self.yCoord))\n",
    "             \n",
    "def main():\n",
    "    ''' Create main function that collects seqname line and prints results'''\n",
    "\n",
    "    data = input('Enter FastQ seqname line')\n",
    "    seqName = FastqString(data)\n",
    "    seqName.parse()\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein coordinates\n",
    "In this exercise, you will create a program that takes three sets of atomic coordinates, all provided on a **single line**. The program then calculates the bond lengths and angles. For this program, you can start with the Triad class (provided). Your task is to create a Python program called coordinateMathSoln that:\n",
    " - asks for and collects three sets of coordinates using input(), only use 1 line for this data !!\n",
    " - outputs the N-C and N-Ca bond lengths and the C-N-Ca bond angle with correct number of significant digits (see below)\n",
    " \n",
    "For example, if I enter the following coordinates (notice.. they are all on one line !!!) :<br>\n",
    "\n",
    "C = (39.447, 94.657, 11.824) N = (39.292, 95.716, 11.027) Ca = (39.462, 97.101, 11.465) <br>\n",
    "\n",
    "then the program will output the following three lines:<br>\n",
    "N-C bond length = 1.33<br>\n",
    "N-Ca bond length = 1.46<br>\n",
    "C-N-Ca bond angle = 124.0<br> \n",
    "(Note: make sure that the angle returned is in **degrees** !!)\n",
    "### Hints: \n",
    " - Each coordinate will contain only 3 numeric values. \n",
    " - Bond lengths are the distance between two points, which for points P and Q in 3-space,\n",
    "$(P_x , P_y , P_z )$ and $(Q_x , Q_y , Q_z )$ respectively, the distance between them is:\n",
    "\n",
    "\n",
    "$$\n",
    "\\begin {eqnarray}\n",
    "     \\|PQ\\| &=&\\sqrt{(P_x-Q_x)^2 + (P_y-Q_y)^2 + (P_z-Q_z)^2)} \\\\\n",
    "            &=&\\sqrt{\\sum^{i \\in x,y,z}{(P_i-Q_i)^2}} \\\\\n",
    "\\end{eqnarray}\n",
    "$$\n",
    "\n",
    " - Bond angles can be calculated from the dot product.\n",
    " \n",
    "Let’s say that we have three points in space labeled P, Q and R. We are interested in the angle at point Q that is made from line segments QP and QR.\n",
    "The dot product tells us (for standard vectors P and R) that:\n",
    "$$P \\cdot R=\\|P\\| \\|R\\| \\cos{\\theta}$$\n",
    "in this notation,$\\|P\\|$ refers to the length of vector P as a standard vector (assumed to begin at the origin (0,0,0) ).\n",
    "We can then see that the angle between vectors P and R can be found by:\n",
    "$$\\cos{\\theta}=\\frac{P \\cdot R} {\\|P\\| \\|R\\|}$$\n",
    "\n",
    "We can calculate the dot product using the sum of products of the vector components:\n",
    "$$P \\cdot R = \\sum^{i \\in x,y,z}{P_i R_i} $$\n",
    "\n",
    "Now, to find vector P in standard form, we need to remember that QP starts at Q, so we need to place the origin at Q and find out where P is in that new space. We do that by subtracting the components of Q from P.\n",
    "Putting all of this together, we get:\n",
    "$$\\theta=\\cos^{-1}{\\frac{\\sum^{i \\in x,y,z}{(P_i - Q_i) (R_i - Q_i) }} {\\|QP\\| \\|QR\\|}}$$\n",
    "Remember, $\\theta$ is in radians.\n",
    "\n",
    "Below I have given you a class (Triad) with methods to calculate dot products (dot), dot products of translated vectors (ndot), distances (dPQ, dPR, dPR) and angles in radians (angleP, angleQ and angleR) for each of the three points in a Triad object.\n",
    "A reasonable solution for this exercise involves around 12 lines of additional code excluding comments.\n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and line comments (using #)\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs as a program level docstring\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nteract": {
     "transient": {
      "deleting": false
     }
    }
   },
   "source": [
    "## Inspection Introduction\n",
    "\n",
    "For this wweeks inspection please focus on coordinateMathSoln. COnsider the design choices that you made, how you are dealing with the dingle line of input data, how you are doing the data conversions, and how you convert to degrees.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### coordinateMathSoln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input protein coordinates in one line: C = (39.447, 94.657, 11.824) N = (39.292, 95.716, 11.027) Ca = (39.462, 97.101, 11.465)\n",
      "N-C bond length = 1.33\n",
      "N-Ca bond length = 1.46\n",
      "C-N-Ca bond angle = 124.0\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Cindy Liang\n",
    "# Group Members: Ben Matera\n",
    "\n",
    "'''\n",
    "Program overview\n",
    "Input: three sets of atomic coordinates, all provided on a single line\n",
    "Output: N-C and N-Ca bond lengths and the C-N-Ca bond angle with correct number of significant digits \n",
    "(3 decimal places for bond lengths and 1 decimal for angle). Angle should be reported in degrees.\n",
    "\n",
    "'''\n",
    "\n",
    "import math\n",
    "class Triad :\n",
    "    \"\"\"\n",
    "    Calculate angles and distances among a triad of points.\n",
    " \n",
    "    Author: David Bernick\n",
    "    Date: March 21, 2013\n",
    "    Points can be supplied in any dimensional space as long as they are consistent.\n",
    "    Points are supplied as tupels in n-dimensions, and there should be three\n",
    "    of those to make the triad. Each point is positionally named as p,q,r\n",
    "    and the corresponding angles are then angleP, angleQ and angleR.\n",
    "    Distances are given by dPQ(), dPR() and dQR()\n",
    " \n",
    "    Required Modules: math\n",
    "    initialized: 3 positional tuples representing Points in n-space\n",
    "             p1 = Triad( p=(1,0,0), q=(0,0,0), r=(0,1,0) )\n",
    "    attributes: p,q,r the 3 tuples representing points in N-space\n",
    "    methods:  angleP(), angleR(), angleQ() angles measured in radians\n",
    "          dPQ(), dPR(), dQR() distances in the same units of p,q,r\n",
    " \n",
    "    \"\"\"\n",
    " \n",
    "    def __init__(self,p,q,r) :\n",
    "        \"\"\" Construct a Triad. \n",
    "        \n",
    "        Example construction:\n",
    "            p1 = Triad( p=(1.,0.,0.), q=(0.,0.,0.), r=(0.,0.,0.) ). \n",
    "        \"\"\"\n",
    "        self.p = p\n",
    "        self.q = q\n",
    "        self.r = r\n",
    "# private helper methods\n",
    "    def d2 (self,a,b) : # calculate squared distance of point a to b\n",
    "        return float(sum((ia-ib)*(ia-ib)  for  ia,ib in zip (a,b)))\n",
    "    \n",
    "    def dot (self,a,b) : # dotProd of standard vectors a,b\n",
    "        return float(sum(ia*ib for ia,ib in zip(a,b)))\n",
    "    \n",
    "    def ndot (self,a,b,c) : # dotProd of vec. a,c standardized to b\n",
    "        return float(sum((ia-ib)*(ic-ib) for ia,ib,ic in zip (a,b,c)))\n",
    "    \n",
    "# calculate lengths(distances) of segments PQ, PR and QR\n",
    "    def dPQ (self):\n",
    "        \"\"\" Provides the distance between point p and point q \"\"\"\n",
    "        return math.sqrt(self.d2(self.p,self.q))\n",
    "    \n",
    "    def dPR (self):\n",
    "        \"\"\" Provides the distance between point p and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.p,self.r))\n",
    "    \n",
    "    def dQR (self):\n",
    "        \"\"\" Provides the distance between point q and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.q,self.r))\n",
    "    \n",
    "    def angleP (self) :\n",
    "        \"\"\" Provides the angle made at point p by segments pq and pr (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.q,self.p,self.r) /   math.sqrt(self.d2(self.q,self.p)*self.d2(self.r,self.p)))\n",
    "    \n",
    "    def angleQ (self) :\n",
    "        \"\"\" Provides the angle made at point q by segments qp and qr (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.p,self.q,self.r) /  math.sqrt(self.d2(self.p,self.q)*self.d2(self.r,self.q)))\n",
    " \n",
    "    def angleR (self) :\n",
    "        \"\"\" Provides the angle made at point r by segments rp and rq (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.p,self.r,self.q) /  math.sqrt(self.d2(self.p,self.r)*self.d2(self.q,self.r)))\n",
    "\n",
    "class AngleConversion :    \n",
    "    '''Coverts angle measurements from radians to degrees'''\n",
    "    def __init__ (self, angle) :\n",
    "        self.radianAngle = angle\n",
    "    def radian2Degrees (self) :\n",
    "        '''Converts angles from radians to degrees'''\n",
    "        degrees = (180/math.pi) * self.radianAngle\n",
    "        return degrees\n",
    "    \n",
    "class CleanInput :\n",
    "    '''Takes coordinate information from input string'''\n",
    "    def __init__ (self, coordinates):\n",
    "        self.coord = coordinates\n",
    "\n",
    "    '''collect coordinates from input and make a list of each set of coords'''\n",
    "    #Assumption: coordinates will C, N, and Ca coords will always be inputted in same order\n",
    "    def makeCoords (self) :\n",
    "        inputList = self.coord\n",
    "        splitList = inputList.split('(')\n",
    "        #split list by ) to separate coordinates from each other\n",
    "        splitList.pop(0)\n",
    "        #remove non-coordinate element of list\n",
    "        #remove non coordinate characters from list\n",
    "                \n",
    "        c = splitList[0].replace(\") N = \",\"\")\n",
    "        n = splitList[1].replace(\") Ca = \",\"\")\n",
    "        ca = splitList[2].replace(\")\",\"\")\n",
    "        cleanedList = [c,n,ca]\n",
    "        return cleanedList\n",
    "\n",
    "class MakeTuple :\n",
    "    '''Converts list of coordinates into a tuple for Triad'''\n",
    "    def __init__ (self, cleanedlist) :\n",
    "        self.coordList = cleanedlist\n",
    "    def makeC (self) :\n",
    "        #turn C coordinates into a tuple\n",
    "        cString = self.coordList[0].split(', ')\n",
    "        cstringTuple = tuple(cString)\n",
    "        cTuple = tuple(map(float, cstringTuple))\n",
    "        return cTuple\n",
    "    def makeN (self):\n",
    "        #turn N coordinates into tuple\n",
    "        nString = self.coordList[1].split(', ')\n",
    "        nstringTuple = tuple(nString)\n",
    "        nTuple = tuple(map(float, nstringTuple))\n",
    "        return nTuple\n",
    "    def makeA (self) :\n",
    "        #make Ca coordinates into tuple\n",
    "        aString = self.coordList[2].split(', ')\n",
    "        astringTuple = tuple(aString)\n",
    "        aTuple = tuple(map(float, astringTuple))\n",
    "        return aTuple\n",
    "    \n",
    "class Announce :\n",
    "    '''Print results to user'''\n",
    "    def __init__(self, ncBL, naBL, cncaAngle):\n",
    "        self.ncBond = ncBL\n",
    "        self.naBond = naBL\n",
    "        self.degAngle = cncaAngle\n",
    "    def printResults(self):\n",
    "        #print results to 3 decimal places for lengths, 1 decimal place for angle\n",
    "        print(\"N-C bond length = {0:1.3g}\".format(self.ncBond ))\n",
    "        print(\"N-Ca bond length = {0:1.3g}\".format(self.naBond) )\n",
    "        print(\"C-N-Ca bond angle = {0:0.1f}\".format(self.degAngle))\n",
    "   \n",
    "    \n",
    "def main():\n",
    "    ''' Collect coordinates from user'''\n",
    "    coord = str(input(\"Input protein coordinates in one line: \"))\n",
    "    cleanedCoords = CleanInput(coord)\n",
    "    coordList = cleanedCoords.makeCoords()\n",
    "    tuples = MakeTuple(coordList)\n",
    "    cCoord = tuples.makeC()\n",
    "    nCoord = tuples.makeN()\n",
    "    aCoord = tuples.makeA()\n",
    "          \n",
    "    '''Using Triad code, calculate N-C, N-Ca bond length and C-N-Ca bond angle in degrees to 3 decimal points.'''\n",
    "    coordMath = Triad(cCoord,nCoord,aCoord)\n",
    "    #pq distance\n",
    "    ncBond = coordMath.dPQ()\n",
    "    #qr distance\n",
    "    naBond = coordMath.dQR()\n",
    "    #q angle\n",
    "    radianAngle = coordMath.angleQ()\n",
    "    #turn angle units to degrees\n",
    "    ConvertAngle = AngleConversion(radianAngle)\n",
    "    degAngle = ConvertAngle.radian2Degrees()\n",
    "    \n",
    "    #print results\n",
    "    printMe = Announce(ncBond, naBond, degAngle)\n",
    "    printMe.printResults()\n",
    "    \n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nteract": {
     "transient": {
      "deleting": false
     }
    }
   },
   "source": [
    "## Inspection Results \n",
    "Group member: Ben Matera\n",
    "Findings:\n",
    "Ben suggested to add a : to my input prompt to more easily differentiate between promot vs. input\n",
    "Response: OK"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Codon tables and amino acid letter converters\n",
    "In this exercise, you will create a program that uses mappings to convert sequence information between different amino acid representations.  This includes the 3-letter codon code (RNA and DNA), the one letter amino acid code and the 3-letter amino acid code.\n",
    "The program will use different dictionaries that represent :\n",
    "codon tables – one for DNA and the other for RNA, and\n",
    "amino acid letter representation converters. \n",
    "\n",
    "Your task is to create a Python program called converter that\n",
    "asks for and collects a single input string using input()\n",
    "parses the string, looks up the information in the appropriate dictionary, and outputs the correct conversion\n",
    "For example:<br>\n",
    "\n",
    "if I enter “ATG” (without quotes), then the program will output:<br>\n",
    "ATG = MET<br>\n",
    " \n",
    "if I enter “UAG” (without quotes), then the program will output:<br>\n",
    "UAG = ---<br>\n",
    "if I enter “E” (without quotes), then the program will output:<br>\n",
    "E = GLU<br>\n",
    "if I enter “Asp” (without quotes), then the program will output:<br>\n",
    "ASP = D<br>\n",
    "### Hints: \n",
    " - The program might not get a valid codon.  In that case, it should output 'unknown'.  You can use the dictionary method 'get' and include a default_value to handle the 'unknown' case.  See Model p. 72 for an example.\n",
    " \n",
    "To get full credit on this assignment, your program needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and line comments (using #)\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs as a program level docstring\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### converter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ASP\n",
      "ASP = D\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Cindy Liang\n",
    "# Group Members: Ben Matera\n",
    "\n",
    "'''\n",
    "Program overview\n",
    "This program uses mappings to convert sequence information between different amino acid representations (3-letter codon in\n",
    "RNA and DNA, 1 letter amino acid code, 3-letter amino acid code. \n",
    "The amino acid representations are stored in dictionaries. \n",
    "\n",
    "input: string\n",
    "output: Correct conversion of string to dictionary entry printed to screen in caps, separated by equal sign. If an invalid\n",
    "codon is given, it will print 'unknown'\n",
    "'''\n",
    "\n",
    "#Create dictionary for 1 letter AA code\n",
    "\n",
    "short_AA = {\n",
    "            'CYS': 'C', 'ASP': 'D', 'SER': 'S', 'GLN': 'Q', 'LYS': 'K',\n",
    "            'ILE': 'I', 'PRO': 'P', 'THR': 'T', 'PHE': 'F', 'ASN': 'N', \n",
    "            'GLY': 'G', 'HIS': 'H', 'LEU': 'L', 'ARG': 'R', 'TRP': 'W', \n",
    "            'ALA': 'A', 'VAL': 'V', 'GLU': 'E', 'TYR': 'Y', 'MET': 'M'\n",
    "            }\n",
    "\n",
    "#3 letter AA code dictionary\n",
    "long_AA = {value:key for key,value in short_AA.items()}\n",
    "\n",
    "#RNA codon dictionary\n",
    "RNA_codon_table = {\n",
    "# Second Base\n",
    "# U             C             A             G\n",
    "#U\n",
    "'UUU': 'Phe', 'UCU': 'Ser', 'UAU': 'Tyr', 'UGU': 'Cys',\n",
    "'UUC': 'Phe', 'UCC': 'Ser', 'UAC': 'Tyr', 'UGC': 'Cys',\n",
    "'UUA': 'Leu', 'UCA': 'Ser', 'UAA': '---', 'UGA': '---',\n",
    "'UUG': 'Leu', 'UCG': 'Ser', 'UAG': '---', 'UGG': 'Trp',\n",
    "#C \n",
    "'CUU': 'Leu', 'CCU': 'Pro', 'CAU': 'His', 'CGU': 'Arg',\n",
    "'CUC': 'Leu', 'CCC': 'Pro', 'CAC': 'His', 'CGC': 'Arg',\n",
    "'CUA': 'Leu', 'CCA': 'Pro', 'CAA': 'Gln', 'CGA': 'Arg',\n",
    "'CUG': 'Leu', 'CCG': 'Pro', 'CAG': 'Gln', 'CGG': 'Arg',\n",
    "#A\n",
    "'AUU': 'Ile', 'ACU': 'Thr', 'AAU': 'Asn', 'AGU': 'Ser',\n",
    "'AUC': 'Ile', 'ACC': 'Thr', 'AAC': 'Asn', 'AGC': 'Ser',\n",
    "'AUA': 'Ile', 'ACA': 'Thr', 'AAA': 'Lys', 'AGA': 'Arg',\n",
    "'AUG': 'Met', 'ACG': 'Thr', 'AAG': 'Lys', 'AGG': 'Arg',\n",
    "#G\n",
    "'GUU': 'Val', 'GCU': 'Ala', 'GAU': 'Asp', 'GGU': 'Gly',\n",
    "'GUC': 'Val', 'GCC': 'Ala', 'GAC': 'Asp', 'GGC': 'Gly',\n",
    "'GUA': 'Val', 'GCA': 'Ala', 'GAA': 'Glu', 'GGA': 'Gly',\n",
    "'GUG': 'Val', 'GCG': 'Ala', 'GAG': 'Glu', 'GGG': 'Gly'\n",
    "}\n",
    "\n",
    "#dna codon dictionary\n",
    "dnaCodonTable = {key.replace('U','T'):value for key, value in RNA_codon_table.items()}\n",
    "\n",
    "'''Collect input from user and see if input corresponds to an entry in any of the dictionaries'''\n",
    "\n",
    "class ReadInput: \n",
    "    def __init__(self, inputString):\n",
    "        self.query = inputString\n",
    "            \n",
    "    def whatAmI(self, inputString):\n",
    "        '''Check if each input is a dictionary entry. if it is none, print unknown'''\n",
    "        \n",
    "        if inputString in short_AA:\n",
    "            myAnswer = short_AA.get(inputString)\n",
    "            print(\"{0} = {1}\".format(self.query, myAnswer.upper()))\n",
    "        elif inputString in long_AA:\n",
    "            myAnswer = long_AA.get(inputString)\n",
    "            print(\"{0} = {1}\".format(self.query, myAnswer.upper()))\n",
    "        elif inputString in RNA_codon_table:\n",
    "            myAnswer = RNA_codon_table.get(inputString)\n",
    "            print(\"{0} = {1}\".format(self.query, myAnswer.upper()))\n",
    "        elif inputString in dnaCodonTable:\n",
    "            myAnswer = dnaCodonTable.get(inputString)\n",
    "            print(\"{0} = {1}\".format(self.query, myAnswer.upper()))\n",
    "        else:\n",
    "            print('unknown')\n",
    "\n",
    "def main():\n",
    "    \n",
    "    query = input()\n",
    "    #collect input from user\n",
    "   \n",
    "    printResult = ReadInput(query)\n",
    "    printResult.whatAmI(query)\n",
    "    #print dictionary entry\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit your code and answers\n",
    " \n",
    "Important: please save your work before logging out of the computer. This will ensure that you have a copy of your work and you will avoid having to redo everything in the  event that something happens with the lab computers. The two recommended solutions (pick one) at this point are to:\n",
    " - Email your code files to yourself\n",
    " - Copy your code files to a pen drive\n",
    " - Save your work on your own computer if you are using Anaconda ( or equiv)\n",
    "\n",
    "The class will be using canvas to submit the source code files created as part of this lab assignment.\n",
    "For this lab, you should upload the following programs as files:\n",
    " \n",
    " - this notebook file, complete with inspection introduction and results for coordMathSoln (4 points)\n",
    " - seqCleaner.py 9 points\n",
    " - fastqParse.py 9 points\n",
    " - coordinateMathSoln.py 9 points\n",
    " - converter.py 9 points\n",
    " - triad.py Triad rewrite (optional) 5 points\n",
    " \n",
    "Important: to get full credit on this lab assignment, each of the code files you submit needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain proper docstrings and appropriate line style comments (#)\n",
    " \n",
    "Congratulations, you finished your second lab assignment!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
