{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Lab 04 Files and Modules\n",
    "\n",
    "# Deliverables:\n",
    " - sequenceAnalysis.py module - 31 total points\n",
    " - classes included: \n",
    "     - NucParams, \n",
    "     - ProteinParams, \n",
    "     - FastAreader\n",
    " - genomeAnalyzer.py - 15 total points\n",
    " - Lab 4 notebook with Inspection Intro Cell and Inspection Results Cell completed( 4 pts)\n",
    " - possible extra credit - 10 additional points\n",
    " - Due: Monday Jan 31, 2022 11:55pm\n",
    "\n",
    "\n",
    "Congratulations. You have started to build an inventory of some pretty useful functions.  Because these are written as classes, you can easily reuse them. Your ProteinParam class is your first deposit into your very own sequenceAnalysis toolkit.  In Python, these toolkits are called modules.\n",
    "\n",
    "We are also going to start using real sequence files.  The fastA format, described here: en.wikipedia.org/wiki/FASTA_format is very convenient to use and fully capable of storing sequences of many types. You will be reading these from an input file for this assignment.\n",
    "\n",
    "## Genomic analysis\n",
    "\n",
    "There are a few things that we can do that mirror and extend the analyses that we did previously on protein sequences. We can calculate composition statistics on the genome (gc content for example), we can calculate relative codon usage in the genome, and we can calculate amino acid composition by translating those codons used the genome.\n",
    "\n",
    "For this lab, I have provided a NucParams class, with the required methods that it implements (see below). You will need to design and write those methods, and these are to be placed in a file called sequenceAnalysis.py This is a __*module*__ that you can use from now on.\n",
    "\n",
    "You will also need to place the ProteinParams class (Lab 3) into this module also. This class will not be used for this assignment, but place it into your toolbox.\n",
    "\n",
    "I have written the FastAreader class.  It is included (below). Keep it as is part of your module for now, you may decide to keep it somewhere else later.\n",
    "\n",
    "The input file for this assignment will be named testGenome.fa, and is available in Canvas. You will not need to submit testGenome.fa, but it will be necessary for your testing.  For development and testing, create a new directory (Lab04) and place the data file (testGenome.fa), your Lab04 notebook, your program (genomeAnalyzer.py) and your new module (sequenceAnalysis.py) into your Lab04 directory.\n",
    "\n",
    "\n",
    "## Hints\n",
    "\n",
    " - Python modules have the .py extension as files, but when they are imported, the name without the extension is used in the import statement in your program.\n",
    "\n",
    " - File placement: Make sure to place your notebook, program, sequenceAnalysis module and the required data files in the same folder. This will allow Python to find them. Read over the FastAreader usage to see how to specify file names that you can use for your data.\n",
    "\n",
    "## Codon frequency calculations\n",
    "\n",
    "Notice that NucParams does all of the counting you need. It is responsible for counts of codons and their translated amino acids.\n",
    "\n",
    "Your genomeAnalyzer.py program has the task of determining which codons are preferred for each of the amino acids and calculating the relative percentage.  For any given amino acid, the relative codon usage (percentages) should sum to 100.0%. Notice that Methionine and Tryptophan only have 1 codon that codes for those, so these will have relative codon usages of 100%.\n",
    "\n",
    "For example: Lysine is coded by both AAA (607) and AAG (917) (example counts in parentheses).  From our aaComposition() method, we are given the aaComposition dictionary and we can lookup 'K' to find 1524 counts (these came from those 607+917 codons).  We can then calculate 607/1524 for AAA and 917/1524 for AAG.  The associated percentages are thus: 39.8 for AAA and 60.2 for AAG.\n",
    "\n",
    "AAA = 607/1524 * 100 = 39.8%\n",
    "\n",
    "AAG = 917/1524 * 100 = 60.2%\n",
    "\n",
    "\n",
    "## Design specification - sequenceAnalysis.py\n",
    "\n",
    "### \\_\\_init\\_\\_\n",
    "\n",
    "The constructor of the class has one optional parameter, a sequence of type string. It may include upper or lower case letters of the set {ACGTUN} or whitespace.  These will be gene sequences and they begin in frame 1.  In other words the first 3 letters of the sequence encode the first AA of the sequence. Carefully consider in what form this class should maintain its data. Is a string the best structure? This class (NucParams) is intended to be very similar to ProteinParam. Make sure to read addSequence() before making this decision, and remember that objects of this class may need to handle an arbitrarily large number of sequences (hint:  dictionaries are good). As a second hint, notice that __init__ and addSequence are doing VERY similar things - you could just make one of them do most of the work.\n",
    "\n",
    "### addSequence() - 5 pts\n",
    "\n",
    "This method must accept new sequences, from the {ACGTUN} alphabet, and can be presumed to start in frame 1. This data must be added to the data that you were given with the __init__ method (if any).\n",
    "\n",
    "### aaComposition() - 6 pts\n",
    "\n",
    "This method will return a dictionary of counts over the 20 amino acids and stop codons.  This dictionary is VERY similar to the lab 3 aaComposition, though you must decode the codon first. The translation table from codon to AA is provided. You are counting amino acids by translating from the proper codon table.\n",
    "\n",
    "### nucComposition() - 10 pts\n",
    "\n",
    "This method returns a dictionary of counts of valid nucleotides found in the analysis. (ACGTNU}. If you were given RNA nucleotides, they should be counted as RNA nucleotides. If you were given DNA nucleotides, they should be counted as DNA nucleotides. Any N bases found should be counted also. Invalid bases are to be ignored in this dictionary.\n",
    "\n",
    "### codonComposition() - 10 pts\n",
    "\n",
    "This dictionary returns counts of codons. Presume that sequences start in frame 1, accept the alphabet {ACGTUN} and store codons in RNA format, along with their counts. __Any codons found with invalid bases should be discarded__. Discard codons that contain N bases. This means that all codon counts are stored as RNA codons, even if the input happens to be DNA. If you discard a codon, make sure to not alter the frame of subsequent codons.\n",
    "\n",
    "### nucCount()\n",
    "\n",
    "This returns an integer value, summing every valid nucleotide {ACGTUN} found.  This value should exactly equal the sum over the nucleotide composition dictionary."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Design specification - genomeAnalyzer.py\n",
    "\n",
    "This program must import your sequenceAnalysis module.\n",
    "It is responsible for preparing the summaries and final display of the data.\n",
    "\n",
    "## Input must be from STDIN\n",
    "Your FastaReader object will read data from sys.stdin if it is not given a filename. You can specify a filename for your testing in jupyter and then remove that filename argument when you move to using standard files.\n",
    "\n",
    "You would replace:\n",
    "\n",
    "myReader = FastAreader('testGenome.fa') # make sure to change this in order to use stdin\n",
    "\n",
    "with:\n",
    "\n",
    "myReader = FastAreader() # make sure to change this in order to use stdin\n",
    "\n",
    "\n",
    "### Output format - 15 pts\n",
    "\n",
    "The function to output the results of your analysis has specific formatting rules that you must follow to get full credit. These rules are as follows:\n",
    "\n",
    " - First line: sequence length = X.XX Mb with two digits after the decimal and labeled Mb (you need to calculate the number of bases in Mb).\n",
    " - second line is blank\n",
    " - third line: GC content = XX.X% as a percentage with one digit after the decimal\n",
    " - fourth line is blank\n",
    " - lines 5 - 68 are the output statistics on relative codon usage for each codon ordered by codon within each amino acid group as follows:"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "XXX : A F (D)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "where XXX is the three letters for an RNA codon, A is the 1-letter amino acid code, F is relative codon frequency, use {:5.1f} for the format, and D is for codon count, use the format {:6d}. There is a single space between each of these fields.\n",
    "For example ( this is not representative of any real genome ):"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sequence length = 3.14 Mb\n",
    "\n",
    "GC content = 60.2%\n",
    "\n",
    "UAA : -  32.6 (  1041)\n",
    "UAG : -  38.6 (  1230)\n",
    "UGA : -  28.8 (   918)\n",
    "GCA : A  14.1 ( 10605)\n",
    "GCC : A  40.5 ( 30524)\n",
    "GCG : A  30.5 ( 22991)\n",
    "GCU : A  14.9 ( 11238)\n",
    "UGC : C  67.2 (  4653)\n",
    "UGU : C  32.8 (  2270)\n",
    "\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output). \n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - contain proper docstrings for the program, classes, modules and any public functions.\n",
    " - Contain in-line comments\n",
    " \n",
    "## Extra credit - 10 pts possible\n",
    "\n",
    "You now have a very powerful set of classes for evaluating genomes. Write a compareGenomes.py program that compares GC content, aaComposition and relative codon bias of 2 genomes. You will have a halophile genome and a hyperthermophile genome to compare.\n",
    "\n",
    "Submit your code using canvas\n",
    "\n",
    "Congratulations, you have finished your fourth lab assignment!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nteract": {
     "transient": {
      "deleting": false
     }
    }
   },
   "source": [
    "## Inspection Intro"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Inspection info\n",
    "Describe all of the information that your inspection team needs to know to understand your design and implementation. \n",
    "\n",
    "Nuc Params: Save counts of codons, amino acids, and proteins as dictionaries with keys as the codons/amino acids/proteins and values as the counts\n",
    "\n",
    "genome analyser.py: does math on the values of the dictionaries from nuc params and prints results\n",
    "\n",
    "fastareader.py: Dr. B's code that takes in fasta files and returns the header and sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# sequenceAnalysis.py #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "# Name: Cindy liang, celiang\n",
    "# Group Members: Amanda Carbajal, Ben Metera\n",
    "\n",
    "'''Project description\n",
    "input: string(s) from a FASTA file representing nucleotide sequences\n",
    "output: strings that describe the nucleotide composition, codon counts, and counts of the codons' translated amino acids\n",
    "for the sequence(s) entered.'''\n",
    "\n",
    "import sys\n",
    "\n",
    "class NucParams:\n",
    "'''Take input sequence(s) and calculate codon, amino acid, and nucleotide counts for the sequence.'''\n",
    "    \n",
    "    aaComp = dict.fromkeys(['A', 'G', 'T', 'N', 'M', 'S', 'C', 'H', 'D',\n",
    "                           'I', 'P', 'V', 'E', 'K', 'Q', 'W', 'F', 'L', 'R', 'Y'], 0)\n",
    "    #initialize dictionary of all possible amino acids.\n",
    "            \n",
    "    nucComp = dict.fromkeys(['A', 'C', 'G', 'T', 'U', 'N'], 0)\n",
    "    #initialize dictionary with possible nucleic acid symbols in FASTA\n",
    "    \n",
    "    rnaCodonTable = {\n",
    "    # initialize RNA codon table (Dr. B's code)\n",
    "    # U\n",
    "    'UUU': 'F', 'UCU': 'S', 'UAU': 'Y', 'UGU': 'C',  # UxU\n",
    "    'UUC': 'F', 'UCC': 'S', 'UAC': 'Y', 'UGC': 'C',  # UxC\n",
    "    'UUA': 'L', 'UCA': 'S', 'UAA': '-', 'UGA': '-',  # UxA\n",
    "    'UUG': 'L', 'UCG': 'S', 'UAG': '-', 'UGG': 'W',  # UxG\n",
    "    # C\n",
    "    'CUU': 'L', 'CCU': 'P', 'CAU': 'H', 'CGU': 'R',  # CxU\n",
    "    'CUC': 'L', 'CCC': 'P', 'CAC': 'H', 'CGC': 'R',  # CxC\n",
    "    'CUA': 'L', 'CCA': 'P', 'CAA': 'Q', 'CGA': 'R',  # CxA\n",
    "    'CUG': 'L', 'CCG': 'P', 'CAG': 'Q', 'CGG': 'R',  # CxG\n",
    "    # A\n",
    "    'AUU': 'I', 'ACU': 'T', 'AAU': 'N', 'AGU': 'S',  # AxU\n",
    "    'AUC': 'I', 'ACC': 'T', 'AAC': 'N', 'AGC': 'S',  # AxC\n",
    "    'AUA': 'I', 'ACA': 'T', 'AAA': 'K', 'AGA': 'R',  # AxA\n",
    "    'AUG': 'M', 'ACG': 'T', 'AAG': 'K', 'AGG': 'R',  # AxG\n",
    "    # G\n",
    "    'GUU': 'V', 'GCU': 'A', 'GAU': 'D', 'GGU': 'G',  # GxU\n",
    "    'GUC': 'V', 'GCC': 'A', 'GAC': 'D', 'GGC': 'G',  # GxC\n",
    "    'GUA': 'V', 'GCA': 'A', 'GAA': 'E', 'GGA': 'G',  # GxA\n",
    "    'GUG': 'V', 'GCG': 'A', 'GAG': 'E', 'GGG': 'G'  # GxG\n",
    "    }\n",
    "    \n",
    "    dnaCodonTable = {key.replace('U','T'):value for key, value in rnaCodonTable.items()}\n",
    "    #initialize dna codon dictionary by replacing Us with Ts for all keys in rna codon dictionary\n",
    "\n",
    "    def __init__ (self, inSeq=''):\n",
    "        '''Create init method with optional parameter string.'''\n",
    "        \n",
    "        self.inSeq = input(\"enter sequence here\")\n",
    "        #grab sequence from user\n",
    "        \n",
    "        cleanedString = self.string.replace(' ', '').upper()\n",
    "        #remove whitespaces and capitalize all nucleotides in sequence\n",
    "                \n",
    "        return cleanedString\n",
    "        \n",
    "    def addSequence (self, thisSequence):\n",
    "        '''Accept new sequences, from the {ACGTUN} alphabet starting from frame 1. \n",
    "        Add data to data from init method (if any).'''\n",
    "                \n",
    "        self.thisSequence = self.__init__()\n",
    "        #grab another cleaned string from user\n",
    "        \n",
    "        while thisSequence != '':\n",
    "            #so long as sequences keep being added, clean the sequences and append them to existing string of seqs\n",
    "#         foo =   ''.join([foo, 'ooo'])\n",
    "            self.cleanedString = ''.join([thisSequence, self.cleanedString])\n",
    "        return self.cleanedString\n",
    "\n",
    "    def aaComposition(self):\n",
    "        '''obtain counts of each amino acid in sequence'''\n",
    "        \n",
    "        for key in self.codonTally().keys():\n",
    "            if key in self.aaDict.keys():\n",
    "                self.aaComp[key] += 1\n",
    "        return self.aaComp\n",
    "    \n",
    "    \n",
    "    def nucComposition(self):\n",
    "        '''tally up the number of times a {ACGTNU} nucleic acid appears in final sequence and put number into dictionary \n",
    "        with the nucleic acids as keys. Return the dictionary of counts'''\n",
    "        \n",
    "        for character in self.addSequence.():\n",
    "            if character in nucComp.keys():\n",
    "                self.nucComp[character] += 1\n",
    "\n",
    "        return self.nucComp\n",
    "    \n",
    "    def codonComposition(self):\n",
    "        '''Count the number of times each valid codon appears in sequence. Discard \n",
    "        (assume discard = don't count) codons with invalid nucleotides. Return dictionary of codons : counts'''\n",
    "        \n",
    "        codonTally = dict.fromkeys(list(self.rnaCodonTable.keys()), 0)\n",
    "        #initialize dictionary to store counts for each codon\n",
    "        \n",
    "        rnaSeq = {self.addSequence().replace('T', 'U') for character in self.addSequence()}\n",
    "        #convert Ts > U for the final sequence \n",
    "        \n",
    "        codonList = [rnaSeq[i:i+3] for i in range(0, len(s), 3)]\n",
    "        #split sequence into 3 to make a list of codons for counting\n",
    " \n",
    "        for codon in codonList:\n",
    "        #if codon is valid, add 1 to codon count for each type of codon. if invalid (invalid base or N), remove it\n",
    "        #only valid codons should be in the rna codon table. use rne codon table to filter out invalid codons.\n",
    "            if codon in self.rnaCodonTable.keys():                \n",
    "                codonTally[codon] += 1\n",
    "       \n",
    "        return codonTally\n",
    "    \n",
    "    def nucCount(self):\n",
    "        return sum(self.nucComp)\n",
    "\n",
    "\n",
    "class FastAreader :\n",
    "    ''' \n",
    "    Define objects to read FastA files.\n",
    "    \n",
    "    instantiation: \n",
    "    thisReader = FastAreader ('testTiny.fa')\n",
    "    usage:\n",
    "    for head, seq in thisReader.readFasta():\n",
    "        print (head,seq)\n",
    "    '''\n",
    "    def __init__ (self, fname=''):\n",
    "        '''contructor: saves attribute fname '''\n",
    "        self.fname = fname\n",
    "            \n",
    "    def doOpen (self):\n",
    "        ''' Handle file opens, allowing STDIN.'''\n",
    "        if self.fname is '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "        \n",
    "    def readFasta (self):\n",
    "        ''' Read an entire FastA record and return the sequence header/sequence'''\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        \n",
    "        with self.doOpen() as fileH:\n",
    "            \n",
    "            header = ''\n",
    "            sequence = ''\n",
    "            \n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>') :\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            for line in fileH:\n",
    "                if line.startswith ('>'):\n",
    "                    yield header,sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else :\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header,sequence\n",
    "        \n",
    "#!/usr/bin/env python3\n",
    "# Name: Cindy liang, celiang\n",
    "# Group Members: Amanda Carbajal\n",
    "\n",
    "'''Project description\n",
    "input: protein sequence in 1 letter code form, in a string\n",
    "output: strings that describe the # amino acids, total molec weight, molar extinction coefficient, mass extinction coefficient,\n",
    "pI, amino acid composition of the protein\n",
    "\n",
    "'''\n",
    "\n",
    "class ProteinParam :\n",
    "    '''Make class for processing input string and calculating chemical properties of protein'''\n",
    "    \n",
    "# These tables are for calculating:\n",
    "#     molecular weight (aa2mw), along with the mol. weight of H2O (mwH2O)\n",
    "#     absorbance at 280 nm (aa2abs280)\n",
    "#     pKa of positively charged Amino Acids (aa2chargePos)\n",
    "#     pKa of negatively charged Amino acids (aa2chargeNeg)\n",
    "#     and the constants aaNterm and aaCterm for pKa of the respective termini\n",
    "#  Feel free to move these to appropriate methods as you like\n",
    "\n",
    "# As written, these are accessed as class attributes, for example:\n",
    "# ProteinParam.aa2mw['A'] or ProteinParam.mwH2O\n",
    "\n",
    "    aa2mw = {\n",
    "        'A': 89.093,  'G': 75.067,  'M': 149.211, 'S': 105.093, 'C': 121.158,\n",
    "        'H': 155.155, 'N': 132.118, 'T': 119.119, 'D': 133.103, 'I': 131.173,\n",
    "        'P': 115.131, 'V': 117.146, 'E': 147.129, 'K': 146.188, 'Q': 146.145,\n",
    "        'W': 204.225,  'F': 165.189, 'L': 131.173, 'R': 174.201, 'Y': 181.189\n",
    "        }\n",
    "\n",
    "    mwH2O = 18.015\n",
    "    aa2abs280= {'Y':1490, 'W': 5500, 'C': 125}\n",
    "\n",
    "    aa2chargePos = {'K': 10.5, 'R':12.4, 'H':6}\n",
    "    aa2chargeNeg = {'D': 3.86, 'E': 4.25, 'C': 8.33, 'Y': 10}\n",
    "    aaNterm = 9.69\n",
    "    aaCterm = 2.34\n",
    "    \n",
    "    \n",
    "    def __init__ (self, protein):\n",
    "        '''Initiate class with protein object'''\n",
    "        self.inString = protein\n",
    "        \n",
    "    def cleanString (self):\n",
    "        '''Create method that cleans input and returns cleaned input'''\n",
    "        capString = self.inString.upper()\n",
    "        #capitalize input\n",
    "        cleanedString = ''\n",
    "        #initialize cleaned input string\n",
    "        for character in capString:\n",
    "            #if character in input is valid, return it. otherwise, remove it\n",
    "            if character in 'AHPWHNVFMTELSDKRCIQY':\n",
    "                cleanedString += character\n",
    "            else:\n",
    "                cleanedString.replace(character, '')\n",
    "\n",
    "        return cleanedString   \n",
    "\n",
    "\n",
    "    def aaCount (self):\n",
    "        '''Create method that returns single integer count of valid amino acid characters found. Assumption: input string\n",
    "        will have invalid characters and spaces. Assume input is read as caps'''\n",
    "        count = len(self.cleanString())\n",
    "        #count valid amino acids\n",
    "        return count\n",
    "    \n",
    "    def molecularWeight (self):\n",
    "        '''Create method that returns molecular weight of protein'''\n",
    "        mw = 0\n",
    "        #initialize mw with 0\n",
    "        count = self.aaCount()\n",
    "        #grab total number of amino acids (for calculating number of peptide bonds later)\n",
    "        for aa in self.cleanString():\n",
    "            #sum weights of all valid amino acids in protein\n",
    "            if aa in self.aaComposition().keys() and aa in self.aa2mw.keys():\n",
    "                mw += self.aa2mw[aa]\n",
    "            else:\n",
    "                mw += 0\n",
    "            \n",
    "        mw -= self.mwH2O*(count-1)\n",
    "        #subtract weight of H2Os used to form peptide bonds\n",
    "        return mw        \n",
    "\n",
    "    def pI (self):\n",
    "        '''Create method that finds the theoretical pI of protein to 2 decimal places'''\n",
    "        pHs = [(i/100) for i in range(0, 1401)]\n",
    "        #create a list of all pHs from 0.00 to 14.00 to calculate charge from\n",
    "        charges = []\n",
    "        #initialize list of charges\n",
    "        for pH in pHs:\n",
    "            #calculate all charges at pHs in possible pH range and add to charges list\n",
    "            charges.append(abs(self._charge_(pH)))\n",
    "\n",
    "        return pHs[charges.index(min(charges))]\n",
    "        #find and return the pH at which the lowest charge occurs for protein    \n",
    "   \n",
    "    def aaComposition (self) :\n",
    "        '''Create method that returns a dictionary listing the counts of all amino acids in protein'''\n",
    "        \n",
    "        aaCompCounts = { v : self.cleanString().count(v) for v in self.cleanString() }\n",
    "        #count occurrence of each amino acid in protein and enter the {amino acid : count} pair into a dictionary\n",
    "        for character in 'AHPWHNVFMTELSDKRCIQY':\n",
    "            #make {amino acid : 0} entry in dictionary for each amino acid that is not in protein\n",
    "            if character in aaCompCounts.keys():\n",
    "                pass\n",
    "            else:\n",
    "                aaCompCounts[character] = 0\n",
    "\n",
    "        return aaCompCounts\n",
    "        \n",
    "    def _charge_ (self, pH):\n",
    "        '''create method that calculates net protein charge at given pH'''\n",
    "        netCharge = 0\n",
    "        #initialize net charge\n",
    "        netCharge += (pow(10, self.aaNterm) / (pow(10, self.aaNterm) + pow(10, pH)))\n",
    "        #add N terminus charge to net\n",
    "        netCharge -= (pow(10, pH) / (pow(10, self.aaCterm) + pow(10, pH)))\n",
    "        #subtract C terminus charge from net\n",
    "        \n",
    "        for key in self.aaComposition():\n",
    "            #Add or subtract positive and negative amino acid charges from net charge\n",
    "            if key in self.aa2chargePos.keys():\n",
    "                netCharge += ( (self.aaComposition()[key] * pow(10, self.aa2chargePos[key]))  / (pow(10, self.aa2chargePos[key]) + pow(10, pH)))\n",
    "                \n",
    "            elif key in self.aa2chargeNeg.keys():\n",
    "                netCharge -= ((self.aaComposition()[key] * pow(10, pH ) / (pow(10, self.aa2chargeNeg[key]) + pow(10, pH))))                                                                            \n",
    "        return netCharge    \n",
    "    \n",
    "\n",
    "    def molarExtinction (self):\n",
    "        '''Create method that calculates molar extinction of protein'''\n",
    "        aa2abs280= {'Y':1490, 'W': 5500, 'C': 125}\n",
    "        molarExt = 0\n",
    "        #initialize molar extinction\n",
    "        \n",
    "        for key in self.aaComposition().keys():\n",
    "            #calculate individual molar extinction of each amino acid and add to net molar extinction\n",
    "            if key in aa2abs280.keys():\n",
    "                molarExt += (self.aaComposition()[key] * aa2abs280[key])\n",
    "            \n",
    "        return molarExt\n",
    "            \n",
    "\n",
    "    def massExtinction (self):\n",
    "        '''Create method that calculates mass extinction of protein'''\n",
    "        myMW =  self.molecularWeight()\n",
    "        return self.molarExtinction() / myMW if myMW else 0.0\n",
    "        #calculate mass extinction of protein\n",
    "\n",
    "\n",
    "# Please do not modify any of the following.  This will produce a standard output that can be parsed\n",
    "\n",
    "    def printOut():\n",
    "        '''Create function that collects input sequence and prints chemical properties of protein'''\n",
    "        inString = input('protein sequence?')\n",
    "        #collect input sequence\n",
    "        while inString :\n",
    "            #print statements\n",
    "            myParamMaker = ProteinParam(inString)\n",
    "            myAAnumber = myParamMaker.aaCount()\n",
    "            print (\"Number of Amino Acids: {aaNum}\".format(aaNum = myAAnumber))\n",
    "            print (\"Molecular Weight: {:.1f}\".format(myParamMaker.molecularWeight()))\n",
    "            print (\"molar Extinction coefficient: {:.2f}\".format(myParamMaker.molarExtinction()))\n",
    "            print (\"mass Extinction coefficient: {:.2f}\".format(myParamMaker.massExtinction()))\n",
    "            print (\"Theoretical pI: {:.2f}\".format(myParamMaker.pI()))\n",
    "            print (\"Amino acid composition:\")\n",
    "\n",
    "            if myAAnumber == 0 : myAAnumber = 1  # handles the case where no AA are present \n",
    "\n",
    "            for aa,n in sorted(myParamMaker.aaComposition().items(), \n",
    "                               key= lambda item:item[0]):\n",
    "                    print (\"\\t{} = {:.2%}\".format(aa, n/myAAnumber))\n",
    "\n",
    "        inString = input('protein sequence?')\n",
    "\n",
    "# if __name__ == \"__main__\":\n",
    "#     main()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GenomeAnalyzer.py #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''This program must import your sequenceAnalysis module. It is responsible for preparing the summaries \n",
    "and final display of the data.\n",
    "\n",
    "input: dictionaries of codons, nucleic acids, and amino acid counts from nucParams\n",
    "output: statistical values of codons, nucleic acids, and amino acids printed out\n",
    "\n",
    "'''\n",
    "\n",
    "import sequenceAnalysis\n",
    "\n",
    "class GeneAnalyzer:\n",
    "    '''Grab counts from nucParams dommath on them, then give them to main to print out'''\n",
    "           \n",
    "    def __init__(self, bases):\n",
    "        self.myNuc = bases\n",
    "           \n",
    "    def numBases(self, bases):\n",
    "        '''calculate num base pairs of sequence in Mb'''\n",
    "        #1 Mb = 1000000 bp\n",
    "        #assume 1 base pair = 1 base in sequence (assume sequence represents one strand)\n",
    "        mb = bases / 1000000\n",
    "        return mb\n",
    "    \n",
    "    def gcCount(self, bases):\n",
    "        '''calculate ratio of of GC bases in sequence'''\n",
    "        \n",
    "        gcContent = ( (bases.count('C') + bases.count('G')) / len(bases) ) * 100\n",
    "        return gcContent\n",
    "\n",
    "def main (fileName=None):\n",
    "    '''Obtain sequences from fasta, obtain tallies of nucleic acid/amino acid/codons from sequence, then print out\n",
    "    composition information'''\n",
    "    \n",
    "    myReader = FastAreader('testGenome.fa') \n",
    "    #grab sequences from fasta with FastAreader and store header, sequence in myReader\n",
    "        \n",
    "    myNuc = NucParams()\n",
    "    #instantiate Nucparams class\n",
    "    \n",
    "    myGeneAnalyzer = GeneAnalyzer()\n",
    "    #instantiate geneAnalyzer\n",
    "    \n",
    "    for head, seq in myReader.readFasta() :\n",
    "        #run sequence through addSeq to get spaces removed\n",
    "        myNuc.addSequence(seq)\n",
    "    \n",
    "    print ('sequence length = {.2f}').format( (len(myNuc) / 1000000 ) )\n",
    "    #compute seq length in Mb by dividing len of final sequence string by 1 mill. print it to 2 decimals.\n",
    "    \n",
    "    print (\"GC content = {:.1}%\".format(myGeneAnalyzer.gcCount(myNuc))\n",
    "    #find gc content of sequence. print to 1 decimal as a percent\n",
    "            \n",
    "    for codon, aa in sorted(myNuc.rnaCodonTable.items(), key = lambda entry: (entry[1],entry[0])):\n",
    "    # sort codons in alpha order, by Amino Acid\n",
    "           \n",
    "        thisCodonComp = myNuc.codonComposition()\n",
    "        #grab tally of all valid codons\n",
    "           \n",
    "        val = thisCodonComp[codon] / sum(thisCodonComp.values())\n",
    "        #calculate relative codon usage in decimals\n",
    "\n",
    "        print ('{:s} : {:s} {:5.1f} ({:6d})'.format(codon, aa, val*100, thisCodonComp[codon]))\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nteract": {
     "transient": {
      "deleting": false
     }
    }
   },
   "source": [
    "## Inspection Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "amanda\n",
    "need to initialize dictionaries with value 0 or else they will have value None\n",
    "change made: added 0 to my dict.fromkeys() function\n",
    "\n",
    "Ben matera\n",
    "Main should be in genomeAnalysis.py\n",
    "change made: put main in genomeAnalysis.py"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
